pkg/catalog/catalog.go

```diff
@@ -77,25 +93,18 @@ func (c *ContextCatalog) Upsert(plugin *cli.PluginInfo) error {
 	// When inserting the "global" or "k8s" target we should remove any similar plugin using
 	// the "unknown" target and vice versa.
 	if plugin.Target == configtypes.TargetGlobal || plugin.Target == configtypes.TargetK8s {
-		c.deleteOldTargetEntries(PluginNameTarget(plugin.Name, configtypes.TargetUnknown))
+		delete(c.plugins, PluginNameTarget(plugin.Name, configtypes.TargetUnknown))
 	} else if plugin.Target == configtypes.TargetUnknown {
-		// This could be a global plugin or a k8s plugin (but not both), so let's
-		// delete either pre-existing entry from the catalog
-		c.deleteOldTargetEntries(PluginNameTarget(plugin.Name, configtypes.TargetGlobal))
-		c.deleteOldTargetEntries(PluginNameTarget(plugin.Name, configtypes.TargetK8s))
+		// An older plugin binary may not specify a target (through its 'info' command).
+		// Therefore the plugin could be a global plugin or a k8s plugin (but not both).
+		// We need to delete either pre-existing entries from the catalog to avoid having
+		// a double entry.
+		delete(c.plugins, PluginNameTarget(plugin.Name, configtypes.TargetGlobal))
+		delete(c.plugins, PluginNameTarget(plugin.Name, configtypes.TargetK8s))
 	}
-
 	return saveCatalogCache(c.sharedCatalog)
```

Comment:
> I was wondering if we could save the catalog as part of the created unlock() function?
I'm thinking that that is when we know that the caller is done modifying the catalog.

Thanks for the suggestion. I have updated the code to throw some meaningful errors when this happens. Also added some unit tests to capture this.

Reply:
Currently, we do not have a place in the code where we create a `contextCatalog` and then use the same catalog to update operations periodically. But if we do that, we would like to ensure that the file persists as soon as possible because subsequent operations can fail or the user can terminate the process in between. It is always good to persist the file as soon as possible to get consistent behavior.

Also, if the `Update`/`Delete` fails while writing to the file we would like to know it as soon as possible during that function call to return a meaningful error message. If we only write during `Unlock` it would be hard to show meaningful error message to the user as we do not know which operation failed actually.